tensorflow/java/src/main/java/org/tensorflow/Tensor.java

```diff
@@ -26,89 +26,130 @@
    *
    * // Valid: A 3x2 matrix of floats.
    * float[][] matrix = new float[3][2];
-   * Tensor m = Tensor.create(matrix);
+   * Tensor<TFFloat> m = Tensor.create(matrix, TFFloat.class);
    *
    * // Invalid: Will throw an IllegalArgumentException as an arbitrary Object
    * // does not fit into the TensorFlow type system.
-   * Tensor o = Tensor.create(new Object());
+   * Tensor<?> o = Tensor.create(new Object(), ...);
    *
    * // Invalid: Will throw an IllegalArgumentException since there are
    * // a differing number of elements in each row of this 2-D array.
    * int[][] twoD = new int[2][];
    * twoD[0] = new int[1];
    * twoD[1] = new int[2];
-   * Tensor x = Tensor.create(twoD);
+   * Tensor<TFInt32> x = Tensor.create(twoD, TFInt32.class);
    * }</pre>
    *
-   * {@link DataType#STRING} typed Tensors are multi-dimensionary arrays of arbitrary byte sequences
-   * and thus have {@code byte[]} and not {@code String}-valued elements. For example:
+   * {@link types.TFString} typed Tensors are multi-dimensional arrays of arbitrary byte sequences,
+   * so can be initialized from arrays of {@code byte[]} elements. For example:
    *
    * <pre>{@code
-   * // Valid: A DataType.STRING tensor.
-   * Tensor s = Tensor.create(new byte[]{1, 2, 3});
+   * // Valid: A TFString tensor.
+   * Tensor<TFString> s = Tensor.create(new byte[]{1, 2, 3}, TFString.class);
    *
    * // Java Strings will need to be encoded into a byte-sequence.
    * String mystring = "foo";
-   * Tensor s = Tensor.create(mystring.getBytes("UTF-8"));
+   * Tensor<TFString> s = Tensor.create(mystring.getBytes("UTF-8"), TFString.class);
    *
-   * // Valid: Matrix of DataType.STRING tensors.
+   * // Valid: Matrix of TFString tensors.
    * // Each element might have a different length.
    * byte[][][] matrix = new byte[2][2][];
    * matrix[0][0] = "this".getBytes("UTF-8");
    * matrix[0][1] = "is".getBytes("UTF-8");
    * matrix[1][0] = "a".getBytes("UTF-8");
    * matrix[1][1] = "matrix".getBytes("UTF-8");
-   * Tensor m = Tensor.create(matrix);
+   * Tensor<TFString> m = Tensor.create(matrix, TFString.class);
    * }</pre>
    *
+   * @param obj The object to convert to a Tensor<T>. Note that whether the it is compatible with
+   *     the type T is not checked by the type system. For type-safe creation of tensors, use {@link
+   *     op.Tensors}.
+   * @param type The class object representing the type T.
    * @throws IllegalArgumentException if {@code obj} is not compatible with the TensorFlow type
-   *     system, or if obj does not disambiguate between multiple DataTypes. In that case, consider
-   *     using {@link #create(DataType, long[], ByteBuffer)} instead.
+   *     system.
+   * @see org.tensorflow.op.Tensors
    */
-  public static Tensor create(Object obj) {
-    return create(obj, dataTypeOf(obj));
+  public static <T extends TFType> Tensor<T> create(Object obj, Class<T> type) {
```

Comment:
IIUC, this is the factory function to use to get type-checking while the `create(Object)` one will generate runtime errors.

This distinction between when types are specified as `DataType` and when as `Class<T>` is a bit confusing, particularly since this `Tensor` class is providing factory functions for both. What do you think about the following:

- `Tensor.java` will NOT provide compile-time type checking. All the methods do runtime checks and functions use the `DataType` enum for type information
- All the "type-safe" construction is with methods in `Tensors.java`

Subtle, but that way we perhaps more clearly layer the use of `DataType` vs. `Class<T>`?
And then over time we can train folks to use `Tensors` for creating `Tensor` objects.

Reply:
I think we need to discuss this. See my comments about Tensor.java:383.

I worry that the unsafe Tensor interface for creation will always "work", so there will be no impetus for people to switch to the safe Tensors interface.

Reply:
It occurs to me that if the Tensor create operation returns a `Tensor<?>` instead of a `Tensor<T>`, that would give programmers the impetus to use the `Tensors` factory instead. I haven't worked through the implications for all the code that currently uses `Tensor.create`, though.

Reply:
I've explored what it would mean to do this last thing. It turns out there is a lot of code in the test cases that is currently using `Tensor.create` to magically get a `Tensor⟨T⟩` of the right T for the surrounding context (regardless of what is actually in the created tensors, since there is no actual compile-time checking that the object being passed is the right type). If the return type of Tensor.create is changed to Tensor⟨?⟩, all this code has to be fixed by inserting a cast to the appropriate type, e.g., `Tensor⟨TFInt32⟩`. That certainly creates an incentive to switch over to type-safe `Tensors` creators but will be annoying for all the existing tests. Maybe that's the right thing -- people will not switch unless `Tensor.create` becomes painful.

The other option I can think of to incentivize use of the type-safe interface is to deprecate `Tensor.create`.